src/Interpreters/Cache/FileCache.cpp

```diff
@@ -1023,7 +1023,8 @@ void FileCache::loadMetadataForKeys(const fs::path & keys_dir)
                                                                       key_metadata,
                                                                       cache_it);
 
-                    inserted = key_metadata->emplace(offset, std::make_shared<FileSegmentMetadata>(std::move(file_segment))).second;
+                    /// TODO: get rid of lock(), it is redundant because it is loadMetadata().
+                    inserted = key_metadata->lock()->emplace(offset, std::make_shared<FileSegmentMetadata>(std::move(file_segment))).second;
```

Comment:
would it make sense to move `lock()` out of loop to lock only once for each key?

Reply:
No, this place in loadMetadata() on startup is tricky, because keys are loaded in parallel, but one keys is always loaded by one thread and noone can access cache until the loadMetadata() is finished. So here it is save to use key_metadata without any lock as it was before this PR. Moving lock before the loop is not possible, because inside the loop we do `lockCache()` for a short period of time - to check if we fit the limits, which means that if we lock key_metadata before the loop we will get lock order inversion - a false positive error.

I wanted to just make loadMetadata as a friend method to KeyMetadata, but it required to move KeyMetadata class in a different .h file, which will not make visible for you the actual diff I made in this PR as all moved struct will be the diff... But this way I like more. wdyt?

Reply:
I don't mind if you'll add a separate commit with this changes